File: algorithm/baekjoon/codeplus/first-solve/src/main/java/com/keunwon/algorithm/dfs/Problem5639.java
package com.keunwon.algorithm.dfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Problem5639 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        Node rootNode = new Node(Integer.parseInt(br.readLine()));

        String w;
        while ((w = br.readLine()) != null && w.length() != 0) {
            int value = Integer.parseInt(w);
            insertNode(rootNode, value);
        }

        postOrder(rootNode);
    }

    static Node insertNode(Node node, int data) {
        if (node == null) {
            return new Node(data);
        }

        Node currentNode;
        if (node.data > data) {
            currentNode = insertNode(node.left, data);
            node.left = currentNode;
        } else {
            currentNode = insertNode(node.right, data);
            node.right = currentNode;
        }
        return node;
    }

    static void postOrder(Node node) {
        if (node != null) {
            postOrder(node.left);
            postOrder(node.right);
            System.out.println(node.data);
        }
    }

    static class Node {
        int data;
        Node left;
        Node right;

        public Node(int data) {
            this.data = data;
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem5639.kt
package com.keunwon.algorithm.baekjoon

/**
 * Title: 이진 검색 트리
 * Level: 골드-5
 **/
class Problem5639 {
    fun solution(arr: IntArray): String {
        val root = Node(arr[0])

        for (i in 1 until arr.size) insert(root, arr[i])
        return postOrder(root)
    }

    private fun postOrder(node: Node?): String {
        if (node == null) return ""
        return postOrder(node.left) + postOrder(node.right) + "${node.num}\n"
    }

    private fun insert(node: Node, n: Int) {
        if (n < node.num) {
            if (node.left == null) node.left = Node(n)
            else insert(node.left!!, n)
        } else {
            if (node.right == null) node.right = Node(n)
            else insert(node.right!!, n)
        }
    }

    private data class Node(
        var num: Int,
        var left: Node? = null,
        var right: Node? = null,
    )
}

fun main() {
    val arr = mutableListOf<Int>()
    while (true) {
        val n = readLine()
        if (n == null || n.isBlank()) break

        arr.add(n.toInt())
    }
    Problem5639().solution(arr.toIntArray()).also(::print)
}
